h2o-automl/src/main/java/ai/h2o/automl/Leaderboard.java

```diff
@@ -370,16 +393,38 @@ public Model getLeader() {
   public static double[] getOtherMetrics(String other_metric, IcedHashMap<Key<ModelMetrics>, ModelMetrics> leaderboard_set_metrics, Frame leaderboardFrame, Model[] models) {
     double[] other_metrics = new double[models.length];
     int i = 0;
-    for (Model m : models)
-      other_metrics[i++] = ModelMetrics.getMetricFromModelMetric(leaderboard_set_metrics.get(ModelMetrics.buildKey(m, leaderboardFrame)), other_metric);
+    for (Model m : models) {
+      // If leaderboard frame exists, get metrics from there
+      if (leaderboardFrame != null) {
+        other_metrics[i++] = ModelMetrics.getMetricFromModelMetric(leaderboard_set_metrics.get(ModelMetrics.buildKey(m, leaderboardFrame)), other_metric);
+      } else {
+        // otherwise use cross-validation metrics
+        Key model_key = m._key;
+        long model_checksum = m.checksum();
+        Key frame_key = m._output._cross_validation_metrics.frame()._key;
+        long frame_checksum = m._output._cross_validation_metrics.frame().checksum();
+        other_metrics[i++] = ModelMetrics.getMetricFromModelMetric(leaderboard_set_metrics.get(ModelMetrics.buildKey(model_key, model_checksum, frame_key, frame_checksum)), other_metric);
+      }
+    }
     return other_metrics;
   }
 
   public static double[] getSortMetrics(String sort_metric, IcedHashMap<Key<ModelMetrics>, ModelMetrics> leaderboard_set_metrics, Frame leaderboardFrame, Model[] models) {
     double[] sort_metrics = new double[models.length];
     int i = 0;
-    for (Model m : models)
-      sort_metrics[i++] = ModelMetrics.getMetricFromModelMetric(leaderboard_set_metrics.get(ModelMetrics.buildKey(m, leaderboardFrame)), sort_metric);
+    for (Model m : models) {
+      // If leaderboard frame exists, get metrics from there
+      if (leaderboardFrame != null) {
+        sort_metrics[i++] = ModelMetrics.getMetricFromModelMetric(leaderboard_set_metrics.get(ModelMetrics.buildKey(m, leaderboardFrame)), sort_metric);
+      } else {
+        // otherwise use cross-validation metrics
+        Key model_key = m._key;
```

Comment:
This is violating the invariants of ModelMetrics.

Better would be to possible to use ModelMetrics.withModelAndFrame() to derive new metrics instead?

Reply:
Yes, good point. Otherwise you will compromise the ModelMetrics object down the line.

Reply:
Actually it is even simpler. It can be changed to:

```
sort_metrics[i++] = ModelMetrics.getMetricFromModelMetric(leaderboard_set_metrics.get(ModelMetrics.buildKey(m, m._output._cross_validation_metrics.frame())), sort_metric);
```

Reply:
This is how would the method look like:

```
  public static double[] getSortMetrics(String sort_metric, IcedHashMap<Key<ModelMetrics>, ModelMetrics> leaderboard_set_metrics, Frame leaderboardFrame, Model[] models) {
    double[] sort_metrics = new double[models.length];
    for (int i = 0; i < models.length; i++) {
      Model m = models[i];
      // If leaderboard frame exists, get metrics from there; otherwise use cross-validation metrics
      Frame metricFrame = leaderboardFrame != null ? leaderboardFrame : m._output._cross_validation_metrics.frame();
      sort_metrics[i++] = ModelMetrics.getMetricFromModelMetric(leaderboard_set_metrics.get(ModelMetrics.buildKey(m, metricFrame)), sort_metric);
    }
    return sort_metrics;
  }
```

Reply:
@navdeep-G Are functions getSortMetric and getOtherMetric different at all?

Reply:
@rpeck We tried/considered using `ModelMetrics.withModelAndFrame()`, right?  The Stacked Ensemble xval metrics are a bit different than a normal model because they are borrowed/cloned from the metalearner -- the metalearner actually trains on a different dataset (cv preds / levelone data) than the SE (train), so that's why there's a difference here.  The xval metrics are cloned from the metalearner, and then we make an update to the framekey to be the training frame instead of the original frame (cv preds / levelone data).

Reply:
See my previous comment. I think code is doing the same thing without changing API

Reply:
@michalkurka I have to revisit that again. I believe they are the same but can’t recall why there are two diff methods.